### DISCUSSION TITLE
jax.experimental.ode.odeint is extremely slow when used in for loop

### DISCUSSION THREAD

**Discussion by pulkitchhabra19** [jax-ml/jax#13365] [Q&A]
Hello, I have a problem where I need to do the ode integration one step at a time i.e. solve the ode integration for one time step and then do some calculations based on the results then do the odeint for another time step. So, I use a for loop like this:

for t in time_steps:
      calculations
      y = odeint(function, y0, t) # where t has t0 and t1 in first iteration and will have t1 and t2 in second iteration and so on ...
      y0 is replaced with y1;

when implementing with JAX, this is extremely slow. the scipy odeint takes ~600 ms; while JAX odeint takes 50+ seconds. 
My next steps would be to: 
a) replace the normal for loop with JAX for loop;
b) try using diffrax

My questions are: 
a) can you explain why this is slow? 
b) would any of the steps that I plan to take help in any way?
c) If not, what would you recommend?

Thanks a lot.

**Comment by pharringtonp19** [comment]
@pulkitchhabra19 You might consider looking at Diffrax. There is a section in the documentation ([see here](https://docs.kidger.site/diffrax/usage/manual-stepping/)) that shows you how to interactively walk through a solve

**Reply by pulkitchhabra19** [reply]
Thanks a lot for your suggestion. But I did try that as well. In fact,  I tried couple of solvers from diffrax and both of them made my code slower than it was with odeint. Based on the "%timeit" results, I have currently:

1. non jax numpy implementation: ~700 to 800 ms. 
2. JAX array based implementation with scipy odeint: ~50 to 60 seconds
3. JAX array based implementation with diffrax Dopri5: ~50 to 60 seconds
4. JAX array based implementation with diffrax Tsit5: ~45 to 50 seconds

DISCLAIMER: At the moment, JAX implementation does not mean that I wrote the code from scratch with JAX using all features like vmap, jit from JAX but rather that I tried to change my existing numpy based code to JAX by changing numpy arrays to JAX arrays; avoiding indexing wherever possible; using JAX random number generation; and finally using JAX ode solver.

May be until I use those extra features of JAX like vmap, jit, it does not work as efficiently. Also, may be I should try and use the for loop structure of JAX but even that is unclear to me how to use it. So, in short, really looking for some advice here. Thanks a lot.
